src/iosrv/drpc_progress.c

```diff
@@ -272,6 +282,111 @@ destroy_session_node(struct drpc_list *session_node)
 	D_FREE(session_node);
 }
 
+/**
+ * ULT to execute the dRPC handler and send the response back
+ */
+static int
+drpc_handler_ult(void *call_ctx)
+{
+	int			rc;
+	struct drpc_call_ctx	*ctx = (struct drpc_call_ctx *)call_ctx;
+
+	D_INFO("dRPC handler ULT for module=%u method=%u\n",
+	       ctx->call->module, ctx->call->method);
+
+	ctx->session->handler(ctx->call, ctx->resp);
+
+	rc = drpc_send_response(ctx->session, ctx->resp);
+	if (rc != 0)
+		D_ERROR("Failed to send dRPC response (module=%u method=%u)\n",
+			ctx->call->module, ctx->call->method);
+
+	/*
+	 * We are responsible for cleaning up the call ctx.
+	 * Session is a copy - we can free it, but not close it. The main loop
+	 * still manages the sessions.
+	 */
+	drpc_free(ctx->session);
+	drpc_call_free(ctx->call);
+	drpc_response_free(ctx->resp);
+	D_FREE(ctx);
+
+	return rc;
+}
+
+static struct drpc_call_ctx *
+create_call_ctx(struct drpc *session_ctx, Drpc__Call *call,
+		Drpc__Response *resp)
+{
+	struct drpc_call_ctx *call_ctx;
+
+	D_ALLOC_PTR(call_ctx);
+	if (call_ctx == NULL)
+		return NULL;
+
+	/*
+	 * Need to copy the session context to avoid the main loop potentially
+	 * freeing that memory out from under our thread.
+	 */
+	call_ctx->session = drpc_dup(session_ctx);
```

Comment:
[Suggestion] I'd consider a reference-count-based solution myself. This copying appears to work though.

Reply:
Switched to refcounting in the latest round of commits.